### PR TITLE
Update phpstan to max level

### PR DESCRIPTION
**Pull request by VincentLanglet**
Currently: 37 errors left.

### CODE REVIEW

File: src/Controller/CRUDController.php [hunk 1]
```diff
@@ -881,7 +891,10 @@ public function aclAction(Request $request): Response
 
     public function getRequest(): Request
```
Comment (VincentLanglet) [inline src/Controller/CRUDController.php:892]: Are we sure there is always a request or should I add an exception if there is not ?

Why is this method public  BTW ?

File: src/Admin/AdminHelper.php [hunk 2]
```diff
@@ -165,7 +165,7 @@ public function appendFormFieldElement(AdminInterface $admin, object $subject, s
 
             $modelClassName = $this->getModelClassName(
                 $admin,
-                explode('.', preg_replace('#\[\d*?]#', '', $path))
+                explode('.', preg_replace('#\[\d*?]#', '', $path) ?? '')
```
Comment (VincentLanglet) [inline src/Admin/AdminHelper.php:168]: `getModelClassName` does not support empty array, should we add a null check inside `getModelClassName` ?

File: src/Controller/CRUDController.php [hunk 3]
```diff
@@ -901,8 +914,11 @@ final public function configureAdmin(Request $request): void
             ));
         }
 
+        $pool = $this->get('sonata.admin.pool');
+        \assert($pool instanceof Pool);
+
         try {
-            $this->admin = $this->get('sonata.admin.pool')->getAdminByAdminCode($adminCode);
+            $this->admin = $pool->getAdminByAdminCode($adminCode);
```
Comment (VincentLanglet) [inline src/Controller/CRUDController.php:921]: The issue will still be present in https://github.com/sonata-project/SonataAdminBundle/pull/7190

I think we don't have other option than adding a phpdoc.

File: src/Controller/CRUDController.php [hunk 4]
```diff
@@ -1307,7 +1335,7 @@ protected function handleXmlHttpRequestSuccessResponse(Request $request, object
             'result' => 'ok',
             'objectId' => $this->admin->getNormalizedIdentifier($object),
             'objectName' => $this->escapeHtml($this->admin->toString($object)),
-        ], Response::HTTP_OK);
+        ]);
     }
 
     final protected function assertObjectExists(Request $request): void
```
Comment (VincentLanglet) [inline src/Controller/CRUDController.php:1341]: What should I do inside the `checkParentChildAssociation` method ?
- if `$this->admin->getParentAssociationMapping()` is null ?
- if `$propertyAccessor->getValue($object, $propertyPath);` is null ?

**Review by VincentLanglet — COMMENTED** [review]
There is a lot of error with `$input->getArgument('foo')` which can be string|string[]|null.
Psalm auto-detect the type, but phpstan doesn't without a Kernel and an applicationLoader...

Using \assert is reported as redundant by psalm. Any recommendation ?

**Comment by VincentLanglet** [comment]
You opened https://github.com/sonata-project/SonataAdminBundle/issues/6833 @vladyslavstartsev ; can you take a look at this PR ?

**Comment by greg0ire** [comment]
> phpstan doesn't without a Kernel and an applicationLoader...

Let's use https://github.com/sonata-project/SonataAdminBundle/blob/3.x/tests/App/AppKernel.php ? That will mean warming up the cache for this kernel before running Phpstan.

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (jordisala1991) [reply src/Controller/CRUDController.php:892]: do we use this method? Controller methods can include the Request parameter directly and remove this.

Comment (VincentLanglet) [reply src/Controller/CRUDController.php:892]: Including the Request parameter would add a BC-break to all the action which require the $request.
We could create an issue about this. WDYT ?

Comment (jordisala1991) [reply src/Controller/CRUDController.php:892]: I saw some code from master. The request is already on all actions, but it is not passed to all method that need it. I think the BC break will be really low.

**Review by vladyslavstartsev — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
Still 20 errors
```
 ------ ------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Admin/Pool.php                                                                                                             
 ------ ------------------------------------------------------------------------------------------------------------------------------- 
  239    Parameter #1 $class of method Sonata\AdminBundle\Admin\Pool::getAdminByClass() expects class-string, class-string|null given.  
 ------ ------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   src/Block/AdminPreviewBlockService.php                                                                                                          
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------ 
  55     Parameter #1 $view of method Sonata\BlockBundle\Block\Service\AbstractBlockService::renderPrivateResponse() expects string, string|null given.  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ------------------------------------------------------------------------------ 
  Line   src/Command/QuestionableCommand.php                                           
 ------ ------------------------------------------------------------------------------ 
  69     Cannot call method set() on Symfony\Component\Console\Helper\HelperSet|null.  
 ------ ------------------------------------------------------------------------------ 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Controller/CRUDController.php                                                                                                                                                                   
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  897    Method Sonata\AdminBundle\Controller\CRUDController::getRequest() should return Symfony\Component\HttpFoundation\Request but returns Symfony\Component\HttpFoundation\Request|null.                 
  921    Property Sonata\AdminBundle\Controller\CRUDController<T of object>::$admin (Sonata\AdminBundle\Admin\AdminInterface<T of object>) does not accept Sonata\AdminBundle\Admin\AdminInterface<object>.  
  1369   Method Sonata\AdminBundle\Controller\CRUDController::getSelectedTab() should return array(?'_tab' => string) but returns array(?'_tab' => float|int<min, -1>|int<1, max>|string|true).              
  1385   Parameter #1 $propertyPath of class Symfony\Component\PropertyAccess\PropertyPath constructor expects string|Symfony\Component\PropertyAccess\PropertyPath, string|null given.                      
  1391   Parameter #2 $needle of method Sonata\AdminBundle\Controller\CRUDController<T of object>::equalsOrContains() expects object, object|null given.                                                     
  1396   Parameter #1 $object of method Sonata\AdminBundle\Admin\AdminInterface<object>::toString() expects object, object|null given.                                                                       
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/DependencyInjection/Compiler/ExtensionCompilerPass.php                                                                                                                                                               
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  233    Method Sonata\AdminBundle\DependencyInjection\Compiler\ExtensionCompilerPass::flattenExtensionConfiguration() should return array('global' => array<string, array<string, array('priority' => int)>>, 'excludes' =>      
         array<string, array<string, array('priority' => int)>>, 'admins' => array<string, array<string, array('priority' => int)>>, 'implements' => array<string, array<class-string, array('priority' => int)>>, 'extends' =>   
         array<string, array<class-string, array('priority' => int)>>, 'instanceof' => array<string, array<class-string, array('priority' => int)>>, 'uses' => array<string, array<class-string, array('priority' => int)>>) but  
         returns array<string, array<string, array<string, array('priority' => int)>&nonEmpty>>.                                                                                                                                  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------------------------- 
  Line   src/DependencyInjection/Compiler/ModelManagerCompilerPass.php                                        
 ------ ----------------------------------------------------------------------------------------------------- 
  39     Parameter #1 $object_or_string of function is_subclass_of expects object|string, string|null given.  
 ------ ----------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/DependencyInjection/SonataAdminExtension.php                                                                                                                                             
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  87     Parameter #1 $configuration of method Symfony\Component\DependencyInjection\Extension\Extension::processConfiguration() expects Symfony\Component\Config\Definition\ConfigurationInterface,  
         Symfony\Component\Config\Definition\ConfigurationInterface|null given.                                                                                                                       
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Form/DataTransformer/ModelToIdPropertyTransformer.php                                                                                                                                          
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  182    Argument of an invalid type array<array<T of object>|T of object|Traversable<mixed, T of object>>|T of object|Traversable<mixed, T of object> supplied for foreach, only iterables are supported.  
  188    Cannot call method __toString() on class-string|object.                                                                                                                                            
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------------- 
  Line   src/Maker/AdminMaker.php                                                                                  
 ------ ---------------------------------------------------------------------------------------------------------- 
  130    Property Sonata\AdminBundle\Maker\AdminMaker::$modelClassBasename (string) does not accept string|false.  
 ------ ---------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------- 
  Line   src/Route/RoutesCache.php                                                                          
 ------ --------------------------------------------------------------------------------------------------- 
  76     Parameter #1 $variable_representation of function unserialize expects string, string|false given.  
 ------ --------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Security/Handler/AclSecurityHandler.php                                                                                                                                
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  147    Cannot call method getUser() on Symfony\Component\Security\Core\Authentication\Token\TokenInterface|null.                                                                  
  148    Parameter #1 $user of static method Symfony\Component\Security\Acl\Domain\UserSecurityIdentity::fromAccount() expects Symfony\Component\Security\Core\User\UserInterface,  
         string|Stringable|Symfony\Component\Security\Core\User\UserInterface given.                                                                                                
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------- 
  Line   src/Util/AdminObjectAclManipulator.php                                                      
 ------ -------------------------------------------------------------------------------------------- 
  213    Possibly invalid array key type string|Symfony\Component\Security\Core\User\UserInterface.  
  214    Possibly invalid array key type string|Symfony\Component\Security\Core\User\UserInterface.  
 ------ -------------------------------------------------------------------------------------------- 
```

**Comment by VincentLanglet** [comment]
I had, out of nowhere the following code failing
```
public function historyAction(Request $request): Response
    {
        $id = $request->get($this->admin->getIdParameter());
        $object = $this->admin->getObject($id);

        $this->assertObjectExists($request);
        \assert(null !== $object);
```

This was because `$request->get($this->admin->getIdParameter())` was returning null.

In `assertObjectExists` there is the following code
```
while (null !== $admin) {
            $objectId = $request->get($admin->getIdParameter());

            if (null !== $objectId) {
                $adminObject = $admin->getObject($objectId);

                if (null === $adminObject) {
                    throw $this->createNotFoundException(sprintf(
                        'Unable to find %s object with id: %s.',
                        $admin->getClassnameLabel(),
                        $objectId
                    ));
                }
            }

            $admin = $admin->isChild() ? $admin->getParent() : null;
        }
```
Should we keep the null check `if (null !== $objectId) {` or should we throw a not found exception in these case ? @phansys

File: src/Admin/AbstractAdmin.php [hunk 5]
```diff
@@ -2481,21 +2485,23 @@ private function buildTabMenu(string $action, ?AdminInterface $childAdmin = null
...
-        $this->menu->setChildrenAttribute('class', 'nav navbar-nav');
-        $this->menu->setExtra('translation_domain', $this->getTranslationDomain());
+        $menu = $this->getMenuFactory()->createItem('root');
+        $menu->setChildrenAttribute('class', 'nav navbar-nav');
+        $menu->setExtra('translation_domain', $this->getTranslationDomain());
 
         // Prevents BC break with KnpMenuBundle v1.x
-        if (method_exists($this->menu, 'setCurrentUri')) {
-            $this->menu->setCurrentUri($this->getRequest()->getBaseUrl().$this->getRequest()->getPathInfo());
+        if (method_exists($menu, 'setCurrentUri')) {
```
Comment (jordisala1991) [inline src/Admin/AbstractAdmin.php:2493]: maybe we can remove this method exists? we dont support knp menu bundle 1, do we?

File: src/Controller/CRUDController.php [hunk 6]
```diff
@@ -901,8 +927,13 @@ final public function configureAdmin(Request $request): void
             ));
         }
 
+        $pool = $this->get('sonata.admin.pool');
+        \assert($pool instanceof Pool);
+
         try {
-            $this->admin = $this->get('sonata.admin.pool')->getAdminByAdminCode($adminCode);
+            /** @phpstan-var AdminInterface<T> $admin */
+            $admin = $pool->getAdminByAdminCode($adminCode);
```
Comment (jordisala1991) [inline src/Controller/CRUDController.php:935]: maybe we could replace the try catch wit hasAdminByCode if? not sure if the pool has that method

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 7]
```diff
@@ -257,14 +261,30 @@ private function isSubtypeOf(string $type, string $subject, string $class): bool
...
+
                 $subjectReflection = new \ReflectionClass($subject);
 
                 return $classReflection->isSubclassOf($subject) || $subjectReflection->getName() === $classReflection->getName();
             case 'implements':
+                if (!interface_exists($subject)) {
+                    return false;
+                }
+
                 return $classReflection->implementsInterface($subject);
```
Comment (jordisala1991) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:264]: in those cases (except the first one) you could return both expressions : if + current return

File: src/Manipulator/ServicesManipulator.php [hunk 8]
```diff
@@ -59,7 +59,7 @@ public function addResource($serviceId, $modelClass, $adminClass, $controllerNam
         $code = "services:\n";
 
         if (is_file($this->file)) {
-            $code = rtrim(file_get_contents($this->file));
+            $code = rtrim(file_get_contents($this->file) ?: '');
```
Comment (jordisala1991) [inline src/Manipulator/ServicesManipulator.php:62]: maybe better ??

File: tests/Command/GenerateObjectAclCommandTest.php [hunk 9]
```diff
@@ -51,7 +51,7 @@ public function testExecuteWithDeprecatedDoctrineService(): void
         $application = new Application();
         $application->add($command);
 
-        $command = $application->find(GenerateObjectAclCommand::getDefaultName());
+        $command = $application->find('sonata:admin:generate-object-acl');
```
Comment (jordisala1991) [inline tests/Command/GenerateObjectAclCommandTest.php:54]: you can use the static property too

File: src/Controller/CRUDController.php (hunk 6 above)
Comment (VincentLanglet) [reply src/Controller/CRUDController.php:935]: We can let it that way it will be soon changed: https://github.com/sonata-project/SonataAdminBundle/pull/7190

File: src/Manipulator/ServicesManipulator.php (hunk 8 above)
Comment (VincentLanglet) [reply src/Manipulator/ServicesManipulator.php:62]: file_get_contents($this->file) can return `false`, that the value which lead to an error.

File: tests/Command/GenerateObjectAclCommandTest.php (hunk 9 above)
Comment (VincentLanglet) [reply tests/Command/GenerateObjectAclCommandTest.php:54]: The static property is protected.
Here the error is that `getDefaultName()` return type is string|null.

I don't think it's bad to copy paste the name of the command.
This way if someone rename the command, test will fail, and it will avoid a regression.

**Review by jordisala1991 — DISMISSED** [review]

**Comment by phansys** [comment]
>Should we keep the null check if (null !== $objectId) { or should we throw a not found exception in these case ? @phansys

Indeed. Looking that snippet and given the usage we are doing for `$objectId`, I think we should throw an exception.

File: src/Controller/CRUDController.php [hunk 10]
```diff
@@ -1353,10 +1396,22 @@ private function checkParentChildAssociation(Request $request, object $object):
         $parentAdmin = $this->admin->getParent();
         $parentId = $request->get($parentAdmin->getIdParameter());
 
-        $propertyAccessor = PropertyAccess::createPropertyAccessor();
-        $propertyPath = new PropertyPath($this->admin->getParentAssociationMapping());
-
         $parentAdminObject = $parentAdmin->getObject($parentId);
+        if (null === $parentAdminObject) {
+            throw new \RuntimeException(sprintf(
+                'No object was found in the admin "%s" for the id "%s"',
```
Comment (phansys) [inline src/Controller/CRUDController.php:1402]:
```suggestion
                'No object was found in the admin "%s" for the id "%s".',
```

File: src/Controller/CRUDController.php [hunk 11]
```diff
@@ -1353,10 +1396,22 @@ private function checkParentChildAssociation(Request $request, object $object):
...
+                \get_class($parentAdmin),
+                $parentId
+            ));
+        }
+
+        $parentAssociationMapping = $this->admin->getParentAssociationMapping();
+        if (null === $parentAssociationMapping) {
+            throw new \RuntimeException('The admin has no parent association mapping');
```
Comment (phansys) [inline src/Controller/CRUDController.php:1410]:
```suggestion
            throw new \RuntimeException('The admin has no parent association mapping.');
```

File: src/Datagrid/SimplePager.php [hunk 12]
```diff
@@ -70,7 +70,12 @@ public function getCurrentPageResults(): iterable
             return $this->results;
         }
 
-        $results = TraversableToCollection::transform($this->getQuery()->execute());
+        $query = $this->getQuery();
+        if (null === $query) {
+            throw new \LogicException('Uninitialized query');
```
Comment (phansys) [inline src/Datagrid/SimplePager.php:75]:
```suggestion
            throw new \LogicException('Uninitialized query.');
```

File: src/Datagrid/SimplePager.php [hunk 13]
```diff
@@ -83,26 +88,27 @@ public function getCurrentPageResults(): iterable
...
-     * @throws \RuntimeException the query is uninitialized
+     * @throws \LogicException the query is uninitialized
      */
     public function init(): void
     {
-        if (!$this->getQuery()) {
-            throw new \RuntimeException('Uninitialized query');
+        $query = $this->getQuery();
+        if (null === $query) {
+            throw new \LogicException('Uninitialized query');
```
Comment (phansys) [inline src/Datagrid/SimplePager.php:97]:
```suggestion
            throw new \LogicException('Uninitialized query.');
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 14]
```diff
@@ -51,10 +71,13 @@ public function process(ContainerBuilder $container): void
...
         }
 
+        /**
+         * @phpstan-var ExtensionMap $extensionConfig
+         */
         $extensionConfig = $container->getParameter('sonata.admin.extension.map');
         $extensionMap = $this->flattenExtensionConfiguration($extensionConfig);
 
-        foreach ($container->findTaggedServiceIds(TaggedAdminInterface::ADMIN_TAG) as $id => $attributes) {
+        foreach ($container->findTaggedServiceIds(TaggedAdminInterface::ADMIN_TAG) as $id => $_attributes) {
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:80]: I think this way adds an additional call, but is easier to read.
```suggestion
        foreach (array_keys($container->findTaggedServiceIds(TaggedAdminInterface::ADMIN_TAG)) as $id) {
```

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php [hunk 15]
```diff
@@ -214,12 +221,12 @@ private function flattenExtensionConfiguration(array $config): array
...
             $optionsMap = array_intersect_key($options, $extensionMap);
 
-            foreach ($optionsMap as $key => $value) {
-                foreach ($value as $source) {
-                    if (!isset($extensionMap[$key][$source])) {
-                        $extensionMap[$key][$source] = [];
+            foreach ($extensionMap as $key => &$value) {
+                foreach ($optionsMap[$key] as $source) {
+                    if (!isset($value[$source])) {
+                        $value[$source] = [];
```
Comment (phansys) [inline src/DependencyInjection/Compiler/ExtensionCompilerPass.php:227]: IMO, there is no need for this initialization.

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/DependencyInjection/Compiler/ExtensionCompilerPass.php (hunk 14 above)
Comment (VincentLanglet) [reply src/DependencyInjection/Compiler/ExtensionCompilerPass.php]: I don't really like this. It add one extra calls and maybe some day we will need the tags.
We have some compiler pass which uses tags.

I remove all this unused related changes in order to solve them in another PR.

**Review by jordisala1991 — APPROVED** [review]

File: tests/Command/GenerateObjectAclCommandTest.php (hunk 9 above)
Comment (jordisala1991) [reply tests/Command/GenerateObjectAclCommandTest.php:54]: After doing a similar change on sonata media, I think it is better to just use the static property,can you take a look on how I did it?

Comment (VincentLanglet) [reply tests/Command/GenerateObjectAclCommandTest.php:54]: Can you provide me the link ?

Comment (jordisala1991) [reply tests/Command/GenerateObjectAclCommandTest.php:54]: nvm this is the test! I thought it was the command class

**Comment by VincentLanglet** [comment]
Can you take a new look @phansys ? :)

File: src/Manipulator/ServicesManipulator.php (hunk 8 above)
Comment (phansys) [reply src/Manipulator/ServicesManipulator.php:62]: I don't know what the possible values for `$this->file` are, but I think we could make this check more explicit.
`?:` will forward the execution flow to the else part for values like `0`.

Comment (VincentLanglet) [reply src/Manipulator/ServicesManipulator.php]: Done, I throw an exception if false is returned.

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php [hunk 16]
```diff
@@ -136,24 +139,23 @@ public function reverseTransform($value)
...
-    public function transform($value)
+    public function transform($value): array
     {
         $result = [];
 
         if (!$value) {
             return $result;
         }
 
-        $isArray = \is_array($value);
```
Comment (phansys) [inline src/Form/DataTransformer/ModelToIdPropertyTransformer.php:147]: I don't know what is the reason for this change. This variable seemed correct given the number of `is_array()` calls checking against its value.
Could you please explain?

File: tests/Form/AbstractLayoutTestCase.php [hunk 17]
```diff
@@ -98,7 +101,7 @@ final protected function assertMatchesXpath(string $html, string $expression, in
                 1 === $count ? 'once' : $count.' times',
                 1 === $nodeList->length ? 'once' : $nodeList->length.' times',
                 // strip away <root> and </root>
-                substr($dom->saveHTML(), 6, -8)
+                substr($dom->saveHTML() ?: '', 6, -8)
```
Comment (phansys) [inline tests/Form/AbstractLayoutTestCase.php:104]: Couldn't be `0` a valid value?

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 16 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:147]: phpstan or/and psalm doesn't understand correctly if this check is saved in a variable

File: tests/Form/AbstractLayoutTestCase.php (hunk 17 above)
Comment (VincentLanglet) [reply tests/Form/AbstractLayoutTestCase.php:104]: I wanted to handle the `false` value ; any recommendation ?

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 16 above)
Comment (phansys) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:147]: That's sad. Fortunately, this function is optimized by the compiler, but we should take care of cases where that could not be the same.

File: tests/Form/AbstractLayoutTestCase.php (hunk 17 above)
Comment (VincentLanglet) [reply tests/Form/AbstractLayoutTestCase.php:104]: I would say that `0` is not a valid value since it's HTML.

It's always supposed to be`<root>...</root>` since we're stripping away the root tag.

File: src/Form/DataTransformer/ModelToIdPropertyTransformer.php (hunk 16 above)
Comment (VincentLanglet) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:147]: Yes, but since it's optimized by the compiler we don't need for another solution here.

Comment (phansys) [reply src/Form/DataTransformer/ModelToIdPropertyTransformer.php:147]: I understand. I mean similar cases where the variable could be holding a value returned by a non optimized function.

File: tests/Form/AbstractLayoutTestCase.php (hunk 17 above)
Comment (phansys) [reply tests/Form/AbstractLayoutTestCase.php:104]: >I wanted to handle the false value ; any recommendation ?
```php
substr(false !== $dom->saveHTML() ? $dom->saveHTML() : '', 6, -8)
```

Comment (VincentLanglet) [reply tests/Form/AbstractLayoutTestCase.php:104]: Since we're doing substr, it doesn't matter
```
substr('0', 6, -8) === false === substr('', 6, -8)
```

**Review by phansys — APPROVED** [review]
